icicle/appUtils/ntt/kernel_ntt.cu

```diff
@@ -34,26 +34,30 @@ namespace ntt {
   static __global__ void
   reorder_digits_inplace_kernel(E* arr, uint32_t log_size, bool dit, bool is_normalize, S inverse_N)
   {
-    // launch N threads
+    // launch N threads (per batch element)
     // each thread starts from one index and calculates the corresponding group
     // if its index is the smallest number in the group -> do the memory transformation
     //  else --> do nothing
 
-    const uint32_t idx = blockDim.x * blockIdx.x + threadIdx.x;
+    const uint32_t size = 1 << log_size;
+    const uint32_t tid = blockDim.x * blockIdx.x + threadIdx.x;
+    const uint32_t idx = tid % size;
+    const uint32_t batch_idx = tid / size;
+
     uint32_t next_element = idx;
     uint32_t group[MAX_GROUP_SIZE];
-    group[0] = idx;
+    group[0] = next_element + size * batch_idx;
 
     uint32_t i = 1;
     for (; i < MAX_GROUP_SIZE;) {
       next_element = dig_rev(next_element, log_size, dit);
       if (next_element < idx) return; // not handling this group
       if (next_element == idx) break; // calculated whole group
-      group[i++] = next_element;
+      group[i++] = next_element + size * batch_idx;
     }
 
-    if (i == 1) { // single element in group --> nothing to do (except maybe normalize for INTT)
-      if (is_normalize) { arr[idx] = arr[idx] * inverse_N; }
+    if (i == 1) { // single element in group --> nothing to do (except normalize for INTT)
```

Comment:
I think this `if` can be removed and the remainder of the function will deal with `i == 1` case correctly. This might even improve performance a tiny bit due to decreased thread divergence

Reply:
you are right